juju/application.py

```diff
@@ -361,11 +363,57 @@ def update_allocation(self, allocation):
         if charm_url == self.data['charm-url']:
             raise JujuError('already running charm "%s"' % charm_url)
 
+        # Update charm
         await client_facade.AddCharm(
             url=charm_url,
             channel=channel
         )
 
+        # Update resources
+        response = await self.model.charmstore.entity(charm_url,
+                                                      channel=channel)
+        store_resources = response['Meta']['resources']
+
+        request_data = [client.Entity(self.tag)]
+        response = await resources_facade.ListResources(request_data)
+        existing_resources = {
+            resource.name: resource
+            for resource in response.results[0].resources
+        }
+
+        resources_to_update = [
+            resource for resource in store_resources
+            if resource['Name'] not in existing_resources or
+            existing_resources[resource['Name']].origin != 'upload'
```

Comment:
Do we need to worry about cases where the revision hasn't changed, to avoid unnecessary fetches, or will the controller already handle detecting those?

Reply:
I'm gonna punt on this one, sorry. FWIW I didn't see the golang code comparing revisions so we're probably okay.